File: src/Java/BehavioralPatterns/State/VendingMachine/src/IdleState.java
import java.util.List;

public class IdleState extends State {
    public IdleState(VendingMachine vendingMachine) {
        super(vendingMachine);
        System.out.println("IDLE - Wait for product selection");
    }

    @Override
    public void insertMoney(double amount) {
        System.out.println("Please select a product before inserting any money.");
    }

    @Override
    public void selectProduct(String productCode) {
        Product selectedProduct = vendingMachine.getProducts().stream()
                .filter(p -> p.getCode().equals(productCode))
                .findFirst()
                .orElse(null);

        if (selectedProduct == null) {
            System.out.println("The product code: " + productCode + " does not exist.");
            return;
        }

        if (selectedProduct.getStock() == 0) {
            System.out.println("The product code: " + selectedProduct.getCode() + " is out of stock");
            return;
        }

        vendingMachine.setSelectedProductCode(selectedProduct.getCode());
        System.out.println("Product: " + selectedProduct.getCode() + " with price: " + selectedProduct.getPrice() + " selected.");
        vendingMachine.setState(new PaymentState(vendingMachine));
    }

    @Override
    public void dispenseProduct() {
        System.out.println("Select a product first.");
    }

    @Override
    public void cancel() {
        System.out.println("There is no selected product or payment in process to cancel.");
    }

    @Override
    public void refill(List<Product> products) {
        vendingMachine.setProducts(products);
        int totalStock = vendingMachine.getProducts().stream().mapToInt(Product::getStock).sum();
        System.out.println("Total amount of products: " + totalStock);
    }
}

File: src/PHP/BehavioralPatterns/State/VendingMachine/IdleState.php
<?php

namespace BehavioralPatterns\State\VendingMachine;

class IdleState extends State {
    public function __construct(VendingMachine $vendingMachine) {
        parent::__construct($vendingMachine);
        echo "IDLE - Wait for product selection\n";
    }

    public function insertMoney(float $amount): void {
        echo "Please select a product before inserting any money.\n";
    }

    public function selectProduct(string $productCode): void {
        $selectedProduct = $this->vendingMachine->getProducts()[$productCode] ?? null;

        if ($selectedProduct === null || $selectedProduct->getStock() === 0) {
            echo "The product code: $productCode is out of stock\n";
            return;
        }

        $this->vendingMachine->setSelectedProductCode($productCode);
        echo "Product: $productCode with price: " . number_format($selectedProduct->getPrice(), 2) . " selected.\n";
        $this->vendingMachine->setState(new PaymentState($this->vendingMachine));
    }

    public function dispenseProduct(): void {
        echo "Select a product first.\n";
    }

    public function cancel(): void {
        echo "There is no selected product or payment in process to cancel.\n";
    }

    public function refill(array $products): void {
        $this->vendingMachine->setProducts($products);
        echo "Total amount of products: " . array_sum(array_map(fn($product) => $product->getStock(), $products)) . "\n";
    }
}